crates/starknet-server/tests/test_dump_and_load.rs

```diff
@@ -332,4 +346,116 @@ mod dump_and_load_tests {
             panic!("Could not unpack the transaction from {loaded_transaction:?}");
         }
     }
+
+    #[tokio::test]
+    async fn dump_load_dump_load_on_exit() {
+        let dump_file = UniqueAutoDeletableFile::new("dump_load_dump_load_on_exit");
+        let devnet_dump_first = BackgroundDevnet::spawn_with_additional_args(&[
+            "--dump-path",
+            &dump_file.path,
+            "--dump-on",
+            "exit",
+        ])
+        .await
+        .expect("Could not start Devnet");
+
+        devnet_dump_first
+            .post_json("/create_block".into(), Body::from(json!({}).to_string()))
+            .await
+            .unwrap();
+        devnet_dump_first.mint(DUMMY_ADDRESS, DUMMY_AMOUNT).await;
+
+        // ctrl c
+        send_ctrl_c_signal(&devnet_dump_first.process).await;
+        std::thread::sleep(std::time::Duration::from_secs(1));
+
+        let devnet_dump_second = BackgroundDevnet::spawn_with_additional_args(&[
+            "--dump-path",
+            &dump_file.path,
+            "--dump-on",
+            "exit",
+        ])
+        .await
+        .expect("Could not start Devnet");
+        devnet_dump_second
+            .post_json("/create_block".into(), Body::from(json!({}).to_string()))
+            .await
+            .unwrap();
+        devnet_dump_second.mint(DUMMY_ADDRESS, DUMMY_AMOUNT).await;
+
+        // ctrl c
+        send_ctrl_c_signal(&devnet_dump_second.process).await;
+        std::thread::sleep(std::time::Duration::from_secs(1));
+
+        let devnet_dump_third = BackgroundDevnet::spawn_with_additional_args(&[
+            "--dump-path",
+            &dump_file.path,
+            "--dump-on",
+            "exit",
+        ])
+        .await
+        .expect("Could not start Devnet");
+
+        let last_block = &devnet_dump_third
+            .send_custom_rpc("starknet_getBlockWithTxHashes", json!({ "block_id": "latest" }))
+            .await["result"];
+        assert_eq!(last_block["block_number"], 3);
+    }
+
+    #[tokio::test]
+    async fn dump_load_dump_load_on_transaction() {
```

Comment:
What is the difference between this test and the previous one (`dump_load_dump_load_on_exit`)? What is the purpose of dumping on each transaction if we don't check the generated artifacts?

Reply:
Generated artifacts are checked each load, I was testing this mode locally and I realized that between the second and third run, there should be 4 events but there were only 2 of them.

There was a bug during the development of this PR that caused that so I decided to add a new test like this. It was related to dump_event() function.

Reply:
I understand now, dump mode "on transaction" needs to be tested separately, as you did it. But let's reduce the duplication, by introducing a common function which accepts a dump-on string as a parameter.

Reply:
Fixed in a29d86e83832f7d8d310bb0791efd10472a27666